mlos_bench/mlos_bench/tests/tunables/tunable_definition_test.py

```diff
@@ -174,6 +225,103 @@ def test_numerical_tunable_reversed_range(tunable_type: str) -> None:
         Tunable(name=f'test_{tunable_type}', config=config)
 
 
+@pytest.mark.parametrize("tunable_type", ["int", "float"])
+def test_numerical_weights(tunable_type: str) -> None:
+    """
+    Instantiate a numerical tunable with weighted special values.
+    """
+    json_config = f"""
+    {{
+        "type": "{tunable_type}",
+        "range": [0, 100],
+        "special": [0],
+        "weights": [0.1, 0.9],
+        "default": 0
+    }}
+    """
+    config = json.loads(json_config)
+    tunable = Tunable(name='test', config=config)
+    assert tunable.special == [0]
+    assert tunable.weights == [0.1, 0.9]
+
+
+@pytest.mark.parametrize("tunable_type", ["int", "float"])
+def test_numerical_weights_no_specials(tunable_type: str) -> None:
+    """
+    Raise an error if weights are specified but no special values.
+    """
+    json_config = f"""
+    {{
+        "type": "{tunable_type}",
+        "range": [0, 100],
+        "weights": [0.1, 0.9],
+        "default": 0
+    }}
+    """
+    config = json.loads(json_config)
+    with pytest.raises(ValueError):
+        Tunable(name='test', config=config)
+
+
+@pytest.mark.parametrize("tunable_type", ["int", "float"])
+def test_numerical_weights_non_normalized(tunable_type: str) -> None:
+    """
+    Instantiate a numerical tunable with non-normalized weights
+    of the special values.
+    """
+    json_config = f"""
+    {{
+        "type": "{tunable_type}",
+        "range": [0, 100],
+        "special": [-1, 0],
+        "weights": [0, 10, 90],
+        "default": 0
+    }}
+    """
+    config = json.loads(json_config)
+    tunable = Tunable(name='test', config=config)
+    assert tunable.special == [-1, 0]
+    assert tunable.weights == [0, 10, 90]  # Zero weights are ok
```

Comment:
Why are zero weights ok?  That should fail, right?

Also, there are 3 weights and 2 specials?

Reply:
Setting some weight to 0 is a very convenient way of (temporarily) disabling the special value.

Last weight is for the values in range. i.e. in this example, we pick `-1` with probability `0.0`, special value of `0` with probability `0.1`, and uniform values in range `[0, 100]` with probability '0.9'

Reply:
Disabling a certain choice makes some sense, but there's a more generic approach there that I think we should consider longer term.  Prior attempts called these "focus" regions.

Anyways, that part I'm okay with for now, however unaligned arrays like this are really hard to read and interpret without extra context.  Some more self-describing data structure I think would be much better.

Reply:
Here's another option: leave `special` values as is, but make the (new) `weights` property more self describing:

```json
"range": [0, 100],
"specials": [-1, "auto"],
"weights": {
  "!range": .09,
  "0": 0,  // temporarily remove 0 from the range
  "-1": .9, // temporarily really focus on "-1",
  "auto": .01,
},
```

This seems easier to read than an unaligned array and gives us more flexibility later on to extend it to support decays, distributions, etc.

Reply:
I like that - let me look at other distributions available in ConfigSpace, and see what other parameters we might need to pass. I'll push the update shortly

Reply:
After some thought, I've decided to keep weights as an array and add a separate `range_weight` parameter, i.e.
```javascript
"type": "int",
"range": [0, 100],
"special": [-1, 0],
"weights": [0.1, 0],
"range_weight": 0.9,
// etc.
```
I don't expect that we will have that many special values, so array looks nicer and more concise to me. Also, it's easier to typecheck comparing to the `"!range"` special key, and it is closer to the internal representation in Python.

Reply:
I don't think the last point is super important.
Configs, in my opinion, should be generally human readable.  It's up to the tool to handle converting it into something useful to itself.

Reply:
I'm still not a super huge fan of the two arrays that need to be aligned pattern, but I guess it's ok enough for now to move on.

One request:
- instead of "weights" and "range_weights", let's name it "range_weights" and "special_weights" so that it's slightly more self documenting.

Reply:
Yeah, I was thinking the second version of the `weights` as a dict that we discussed in this version of the thread, instead of the first one that was about reformatting `specials`, but that's OK.  I think it should be good enough for now.